python/pyspark/sql/connect/client/core.py

```diff
@@ -688,6 +653,36 @@ def enable_reattachable_execute(self) -> "SparkConnectClient":
         self._use_reattachable_execute = True
         return self
 
+    def register_retry_policy(self, policy: RetryPolicy):
+        """
+        Registers specified policy in the dictionary of known policies.
+
+        To activate it, use set_retry_policies().
+        """
+        if policy.name in self._known_retry_policies:
+            raise ValueError("Already known policy")
+        self._known_retry_policies[policy.name] = policy
+
+    def set_retry_policies(self, policies: List[Union[str, RetryPolicy]]):
+        """
+        Sets list of policies to be used for retries.
+        I.e. set_retry_policies(["DefaultPolicy", "CustomPolicy"]).
+
+        If policy is given as string, the policy object is sourced from previously
+        registered policies.
+        Specifying policy directly bypasses the registration mechanism.
+        """
+        self._retry_policies = [
+            policy if isinstance(policy, RetryPolicy) else self._known_retry_policies[policy]
+            for policy in policies
+        ]
+
+    def get_retry_policies(self) -> List[str]:
+        """
+        Return list of currently used policies, i.e. ["DefaultPolicy", "MyCustomPolicy"].
+        """
+        return [policy.name for policy in self._retry_policies]
```

Comment:
In my opinion the registering of policies by name is still superfluous. Especially that `policy.name` is currently generated only from policy classname, so it's currently not possible to use a difference instance of the same policy with different parameters (e.g. different sleep timings). So I still think that any power-user who wants to customize it would just define it with their own instances of the objects.
I think that now having some policies be registered under a name in known_retry_policies, but the key there being derived only from class name so registering a different instance will overwrite the entry is confusing, and having the set_retry_policies being either a string from known_retry_policies, or an instance (which may not be part of known_retry_policies, so when it's returned by-name by get_retry_policies, the user can't do anything with that name) is also confusing.

WDYT @HyukjinKwon @grundprinzip  ?

Reply:
I continue to not see sense in having the `register_retry_policy` API and see confusion from either having `known_retry_policies` identified by a string which is just a class name, and different instances of the same class name being easily confused, or by specific instance of the object. I see this as unnecessarily convoluted and not intuitive.

I would propose to leave just
```
    def set_retry_policies(self, policies: List[RetryPolicy]):

    def get_retry_policies(self) -> List[RetryPolicy]:
```
and get rid of the `_known_retry_policies`.

I leave it to other interviewers if they disagree and think that the current API makes sense.

Otherwise, LGTM.